gadget/update.go

```diff
@@ -247,6 +294,34 @@ func EnsureLayoutCompatibility(gadgetLayout *LaidOutVolume, diskLayout *OnDiskVo
 		}
 	}
 
+	// check all structures in the layout are present in the gadget, or have a
+	// valid excuse for absence (i.e. mbr or creatable structures at install)
+	for _, gs := range gadgetLayout.LaidOutStructure {
+		// we ignore reasonAbsent here since if there was an extra on disk
+		// structure that didn't match something in the YAML, we would have
+		// caught it above, this loop can only ever identify structures in the
+		// YAML that are not on disk at all
+		if present, _ := onDiskContains(diskLayout.Structure, gs); present {
+			continue
+		}
+
+		// otherwise not present, figure out if it has a valid excuse
+
+		// allow specifically the mbr structure to be missing from ondisk since
+		// it doesn't show up as a device since it's a "raw" thing
+		if gs.Role == "mbr" {
```

Comment:
should this also account for gs.Type == "bare" ? actually this should probably look like:

```
if gs.IsPartition() {
    continue
}
```

Reply:
ah I think you are right and I just realized that this whole time I have been testing with structures that have a partition type and thus are not `bare`, I need to go back and look at this again

Reply:
ok this is ready again, I changed this to check for IsPartition as you suggested and added a unit test about it too

Reply:
also to be clear I think you meant to have `!gs.Partition()` (that's what I went with anyways)(